rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/KeyPairProducer.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * ProActive Parallel Suite(TM):
+ * The Open Source library for parallel and distributed
+ * Workflows & Scheduling, Orchestration, Cloud Automation
+ * and Big Data Analysis on Enterprise Grids & Clouds.
+ *
+ * Copyright (c) 2007 - 2017 ActiveEon
+ * Contact: contact@activeeon.com
+ *
+ * This library is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Affero General Public License
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.resourcemanager.utils;
+
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.log4j.Logger;
+
+
+public class KeyPairProducer {
+
+    private static final Logger LOGGER = Logger.getLogger(KeyPairProducer.class);
+
+    private AtomicInteger numberOfWorkers;
+
+    private BlockingQueue<KeyPair> keyPairs = new LinkedBlockingQueue<>();
+
+    private long loopTimeout = 1000; // in milliseconds
+
+    // for test purpose, e.g TaskLauncherTest
+    public KeyPairProducer() {
+
+    }
+
+    public KeyPairProducer(int numberOfworkers) {
+        // let say we generate 3 times more keys than nodes there are
+        this.numberOfWorkers = new AtomicInteger(numberOfworkers * 3);
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (!Thread.currentThread().isInterrupted()) {
+                    try {
+                        // if we obeserve less than 'numberOfWorkers' elements than we going to
+                        // add 'numberOfWorkers' more
+                        int numberOfKeysToAdd = numberOfWorkers.get() - keyPairs.size();
+                        if (numberOfKeysToAdd > 0) {
+                            for (int i = 0; i < numberOfKeysToAdd; ++i) {
+                                keyPairs.put(generateAndAddKeyPairs());
+                            }
+                        }
+
+                        // sleep 1 second
+                        synchronized (keyPairs) {
+                            keyPairs.wait(loopTimeout);
```

Comment:
This wait is never awaken by a notify. Is that ok?

Reply:
I think we could get rid of any waiting mechanism (besides the blocking queue) by using a bounded LinkedBlockingQueue, so that the thread would automatically block just above at line 67 on the `put` if the queue is already full, and would be awaken automatically when the task launcher takes a key. We can achieve that by creating it with: `LinkedBlockingQueue(int capacity)`

Reply:
Thanks, right, make sense